tests/integration/events.go

```diff
@@ -240,3 +266,84 @@ func waitForMergeEvents(s *state) {
 		}
 	}
 }
+
+// getEventsForUpdate returns a map of expected update events for an update document action.
+//
+// This will take into account any primary documents that are patched as a result
+// of the create or update.
+func getEventsForUpdate(s *state, action UpdateDoc) map[string]struct{} {
+	var collection client.Collection
+	if action.NodeID.HasValue() {
+		collection = s.collections[action.NodeID.Value()][action.CollectionID]
+	} else {
+		collection = s.collections[0][action.CollectionID]
+	}
+
+	docID := s.docIDs[action.CollectionID][action.DocID]
+	def := collection.Definition()
+
+	docMap := make(map[string]any)
+	err := json.Unmarshal([]byte(action.Doc), &docMap)
+	require.NoError(s.t, err)
+
+	expect := make(map[string]struct{})
+	expect[docID.String()] = struct{}{}
+
+	// check for any secondary relation fields that could publish an event
+	for name, value := range docMap {
+		field, ok := def.GetFieldByName(name)
+		if !ok {
+			continue // ignore unknown field
+		}
+		_, ok = field.GetSecondaryRelationField(def)
+		if ok {
+			expect[value.(string)] = struct{}{}
+		}
+	}
+
+	return expect
+}
+
+// getEventsForCreate returns a map of expected update events for a document create action.
+//
+// This will take into account any primary documents that are patched as a result
+// of the create or update.
+func getEventsForCreate(s *state, action CreateDoc) map[string]struct{} {
+	var collection client.Collection
+	if action.NodeID.HasValue() {
+		collection = s.collections[action.NodeID.Value()][action.CollectionID]
+	} else {
+		collection = s.collections[0][action.CollectionID]
+	}
+
+	docs, err := parseCreateDocs(action, collection)
+	require.NoError(s.t, err)
+
+	def := collection.Definition()
+	expect := make(map[string]struct{})
+
+	for _, doc := range docs {
+		expect[doc.ID().String()] = struct{}{}
+
+		// check for any secondary relation fields that could publish an event
+		for f, v := range doc.Values() {
+			field, ok := def.GetFieldByName(f.Name())
+			if !ok {
+				continue // ignore unknown field
+			}
+			_, ok = field.GetSecondaryRelationField(def)
+			if ok {
+				expect[v.Value().(string)] = struct{}{}
+			}
+		}
+	}
+
+	return expect
+}
+
+// getEventsForDelete returns a map of expected update events for a document delete action.
+func getEventsForDelete(docID client.DocID) map[string]struct{} {
+	expect := make(map[string]struct{})
+	expect[docID.String()] = struct{}{}
```

Comment:
todo: Please dont grow the map like this, please either set the length on `make`, or (IMO) better replace this with:
```
return map[string]struct{}{
   docID.String(): {},
}
```
I would probably delete the `getEventsForDelete` entirely and just do this inline where `getEventsForDelete`, it saves readers from having to bounce around the file when reading (I won't push for that as a todo though :)).

Reply:
I dropped the function and created the map like you suggested.